### ISSUE TITLE
Overlapping Model Effects Not Rendering When in Separate Groups

### ISSUE THREAD

**Issue by idleup** [smeighan/xLights#668]
Example:
I have a group called Wreaths (All) and a Group called Wreaths (Left). While Wreaths (Left) has an effect overlapping Wreaths (All) then only Wreaths (Left) will render in playback.

Desired Behavior:
Layer Blending options would apply to overlapping effects on a element whether the overlapping effect exists in a separate layer for that element (current functionality) or a separate group containing that element (proposed).

**Comment by AzGilrock** [comment]
This is more of a global rendering issue that we've discussed several times.  Currently what's rendered on model A does not blend down onto model B.  It doesn't matter if a model is the same model in 2 different groups because a group is its own independent model that contains a merging of all the pixels of the models in the group.  So effects dropped at the group level are treated as if dropped on a completely different model.  Also the Morph option was a carryover from when xLights only supported 2 effects per model.  It allowed you to morph from effect 1 to effect 2.

**Comment by idleup** [comment]
So this does not sound like an easy fix but it is really important (to me at least) because I sequence in Groups, not elements. Nearly every row in my sequence is a group. Not having the ability to blend between models is the real issue/request here based on your explanation. Since all my sequencing is done in groups, not having the ability to morph between them severely handicaps my capabilities (similar to as if layers did not exist for those who model with elements as their rows in the sequencer)

As for the morph option being limited to 2 effects per model, I am still very new to xLights but that limitation is workable for me as most of the time the morph/blending is used for transitions, so I do not want this request for ability to blend between models (e.g., group rows) to be convoluted with a request for morphing more than 2 effects as that is far less important to me.

If making this work would require significant changes that are not possible in the near future can an interim fix be put in place to at least give the same functionality as 'Layered' or 'Averaged' when a channel exists on two "models" at the same time in the sequence? The reason I say channel is because it seems that the channel number is the only unique identifier the system knows to determine if a pixel that exists in two elements is really the same pixel/element...

**Comment by AzGilrock** [comment]
I don't know all the details because Dan is our specialist for all the rendering engine functionality.  I believe I've heard him say it would be easy to turn on having what's already rendered blend with what's coming on top of it but then that potentially messes up everyone's existing sequences that might not want it to work that way.  Plus we don't have a mechanism to specify which type of blending to do like we have for every effect.  I'm also not sure if what you are seeing is true for all effects since some support alpha channel data and some don't.

**Comment by dkulp** [comment]
Turn on the blending onto what's already rendered into the buffer is easy, it's only changing a false to a true in the code.   The two main problems are:
1) Existing sequences will be hugely problematic.   We can go through and at load time, set the bottom most effects to blending None which would be 90% correct.

2) Figuring out when to clear the buffer.   This is the thing I'm kind of stuck on.   Obviously at the start of the "Render All", the entire buffer would get cleared.  That's easy.   But when you adjust a setting, we'd need to figure out what to do.   Easiest would be to clear the buffer for the nodes of that model.  Most correct would be to clear the nodes for that model and all models and groups that need to render before it and after it and render all of those.   That's WAY more complex and would definitely have performance issues. (although it does fix a few other bugs) . It can also spiral into massive amounts of rendering for a single setting change unless we also add abilities to restrict the rendering to JUST the channels that overlap with the given model, but that's another major complexity.

The other complication is what to do about data layers that occur before the Nutcracker layer.   Instead of a clear, do we pull that data in so that nutcracker effects can blend onto them?

Anyway, I'm fairly confident that this is not likely to happen this year.   I'm kind of winding down the major changes for this year so we can concentrate on stabilizing it and then using it to sequence.

**Comment by idleup** [comment]
Dan, I set the values so that I could blend between model groups but having to constantly hit Render All after every effect change because cumbersome and extremely time consuming so I set it back and started copying the same effect to a new layer on every instance of the model effect I want blended (very inefficient and resulted in many extra layer rows). 

I think the first issue you mention can be resolved by either doing what you said with setting the bottom effect to None and/or by making blending between models an option that the user can turn on at the sequence settings level so existing sequences would retain their existing functionality.

As for the second issue, I think it would be solved by implementing the background render that only renders the changed effect and any overlapping effects that occur during the same time-slice of the changed effect. I logged this as issue #728.
